.github/workflows/tests.yml

```diff
@@ -1,46 +1,222 @@
+    - name: Install/Upgrade black
+      run: python -m pip install --user --upgrade black
+    - name: Check code formatting with black
+      # The --include flag should be removed once all files are
+      # formatted with black. But as long as this is not the case, only
+      # check those files that are already formatted.
+      run: |
+        python -m black ./ --check --line-length 79 --include \
+        setup.py \
+        mdtools/__init__.py \
+        mdtools/_metadata.py \
+        mdtools/version.py
+    - name: Install/Upgrade flake8 and flake8-docstrings extension
+      run: python -m pip install --user --upgrade flake8 flake8-docstrings
+    - name: Lint code and docstrings with flake8
+      # Use setup.cfg to configure flake8
+      # The --select flag should be removed once all files are revised
+      # to be compliant with flake8. But as long as this is not the
+      # case, select only those error codes that have been fixed in all
+      # files.
+      run: |
+        python -m flake8 ./ --select=\
+        E1,E2,E3,E4,E9,\
+        F4,F7,F63,F82,\
+        W1,W2,W3
 
-    runs-on: ubuntu-latest
+  test:
+    # Only test after formatting and linting, because there is a very
+    # small chance that code formatting introduces errors into the code.
+    needs: lint
     strategy:
-      fail-fast: true
       matrix:
+        # Tests must be run on all target platforms and Python versions
+        os: [ubuntu-latest, macos-latest, windows-latest]
         python-version: [3.6, 3.7, 3.8]
-
+      # Do not cancel in-progress jobs if any matrix job fails
+      fail-fast: false
+    runs-on: ${{ matrix.os }}
     steps:
-    - uses: actions/checkout@v2
+    - name: Checkout repository
+      uses: actions/checkout@v2
     - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python-version }}
-    - name: Install dependencies
+    - name: Export ${HOME}/.local/bin to ${PATH}
+      # Executable Python binaries are ususally stored there
+      run: echo "${HOME}/.local/bin" >> ${GITHUB_PATH}
+    - name: Get pip cache dir
+      # pip's cache path depends on the operating system, see
+      # https://github.com/actions/cache/blob/main/examples.md#python---pip
+      # This requires pip >=20.1
+      id: pip-cache
       run: |
-        python -m pip install --upgrade pip
-        python -m pip install flake8 flake8-docstrings pytest
-        python -m pip install numpy
-        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
-    - name: Lint code with flake8
+        python -m pip install --user --upgrade pip
+        echo "::set-output name=dir::$(pip cache dir)"
+    - name: Create/Restore Cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          ${{ steps.pip-cache.outputs.dir }}/**
+          ./docs/build/**
+          ~/docs/build/**
+          ./.pytest_cache/**
+          ~/.pytest_cache/**
```

Comment:
Does caching of build and test artifacts speeds up the build significantly? I would expect that downloading with pip and installing dependencies is the most significant hurdle. Caching artifacts might lead to some unexpected results. I would prefer a clean build if this is not much slower.

Reply:
I don't know what effect caching artifacts of pytest has on speed. So I can't judge this.

But building the docs from scratch takes significantly more time than only building the files that have changed (what is what Sphinx does when you simply do `make -C ./docs html`). If you build from scratch, Sphinx has for instance to download the inventories of the docs of all packages that we reference in our docs (see [here](https://github.com/andthum/mdtools/blob/66217f76915efd79bd84930b06e5ded1bbb11fa8/docs/source/conf.py#L381) for a list of inventories that have to be downloaded). So, on my machine I'm on the order of tens of seconds to minutes if I build the docs from scratch and on the order of a few seconds (or even miliseconds) if I simply call `make -C ./docs html` without doing `make -C ./docs clean clean_autosum` beforehand.

Generally, you are right and I would also prefer a clean build and test environment. But we have only a [limited number of free minutes](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions) to run GitHub Actions (2000 minutes per month). I have no experience, whether this is more than enough or whether we have to be economical. You also have to consider, that every Windows minute counts as 2 minutes and every MacOS minute counts even as 10 minutes.

Another reason to cache and speed up things is that you don't want to wait long for your tests to pass.

But I'm wondering a bit about the available storage for caching. For a free account we should only have 500 MB of [storage for artifacts](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions), but it seems that everyone has 5 GB of [cache storage](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy). If you watch the cache sizes in the "Create/Restore Cache" steps, the 5 GB sum up quickly.

Reply:
I guess it's best then to simply cache everything and deal with problems later. Problems with caching would manifest in failing actions, I assume.